### PR TITLE
Containerd metricbeat module

### PR DESCRIPTION
**Pull request by MichaelKatsoulis**
## What does this PR do?

This PR creates containerd metricbeat module. 
Cpu, memory and blkio metricsets are created.

## Why is it important?

Containerd is a container runtime that implements Container Runtime Interface (CRI).
It is used as one of Kubernetes runtimes after k8s deprecating docker  after v1.20.
Containerd if configured to expose metrics it provides useful informations about cpu, memory and blkio.

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

1. Create a kind kubernetes cluster of version higher than 1.20 
`kind create cluster --image 'kindest/node:v1.21.1' --config kind-cluster.yaml`
2. docker exec in the kind docker container that is created
3. Edit `/etc/containerd/config.toml ` and add 
```
[metrics]
        address = "127.0.0.1:1338"
```
4. Restart containerd service ` systemctl restart containerd`
5. Before deploying metricbeat add the following data in `metricbeat-daemonset-modules` ConfigMap
```
containerd.yml: |-
    - module: containerd
      metricsets:
        - cpu
        - memory
        - blkio
      enabled: true
      period: 10s
      hosts: ["localhost:1338"]
      calcpct: true
```
6. Run Metricbeat and watch the `containerd` fields getting populated.

## Screenshots

containerd_cpu
containerd_memory

containerd_blkio

### CODE REVIEW

File: metricbeat/module/containerd/cpu/cpu.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	mapping := &prometheus.MetricsMapping{
+		Metrics: map[string]prometheus.MetricMap{
+			"container_cpu_total_nanoseconds":  prometheus.Metric("usage.total.ns"),
+			"container_cpu_user_nanoseconds":   prometheus.Metric("usage.user.ns"),
+			"container_cpu_kernel_nanoseconds": prometheus.Metric("usage.kernel.ns"),
+			"container_per_cpu_nanoseconds":    prometheus.Metric("usage.percpu.ns"),
+			"process_cpu_seconds_total":        prometheus.Metric("system.total"),
+		},
+		Labels: map[string]prometheus.LabelMap{
+			"container_id": p.KeyLabel("id"),
```
Comment (ChrsMark) [inline metricbeat/module/containerd/cpu/cpu.go:53]: You are using openmetrics library here, is this intentional? Why not `prometheus.KeyLabel()`?

File: metricbeat/module/containerd/cpu/cpu.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+	// HostParser validates Prometheus URLs
+	hostParser = parse.URLHostParserBuilder{
+		DefaultScheme: defaultScheme,
+		DefaultPath:   defaultPath,
+	}.Build()
+)
+
+// init registers the MetricSet with the central registry.
+// The New method will be called after the setup of the module and before starting to fetch data
+func init() {
```
Comment (ChrsMark) [inline metricbeat/module/containerd/cpu/cpu.go:42]: This is a new module so I guess it would better fit under `xpack`.

File: metricbeat/module/containerd/cpu/metricset.go [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+		}
+		// calculate system usage delta
+		systemUsageDelta := float64(systemTotalNs) - m.preSystemCpuUsage
+
+		// Calculate cpu total usage percentage
+		cpuUsageTotal, err := event.GetValue("usage.total.ns")
+		if err == nil {
+			cpuUsageTotalPct := calcCpuTotalUsagePct(cpuUsageTotal.(float64), systemUsageDelta,
+				float64(contCpus), cID, m.preContainerCpuTotalUsage)
+			m.Logger().Infof("cpuUsageTotalPct for %+v is %+v", cID, cpuUsageTotalPct)
```
Comment (ChrsMark) [inline metricbeat/module/containerd/cpu/metricset.go:130]: Maybe that's too noisy?

File: metricbeat/module/containerd/cpu/metricset.go [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+// Fetch gathers information from the containerd and reports events with this information.
+func (m *metricset) Fetch(reporter mb.ReporterV2) error {
+	events, err := m.prometheusClient.GetProcessedMetrics(m.prometheusMappings)
+	if err != nil {
+		return errors.Wrap(err, "error getting metrics")
+	}
+	var systemTotalNs int64
+	perContainerCpus := make(map[string]int)
+	elToDel := -1
+	for i, event := range events {
```
Comment (ChrsMark) [inline metricbeat/module/containerd/cpu/metricset.go:73]: The additional percentances calculations can be configured to be enabled/disabled on demand like what is happening in docker module so as to avoid overloading the in case we have too many containers/metrics and we are interested into getting that much detail.

File: metricbeat/module/containerd/cpu/metricset.go [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+		}
+		contCpus, ok := perContainerCpus[cID]
+		if !ok {
+			contCpus = 1
+		}
+		// calculate system usage delta
+		systemUsageDelta := float64(systemTotalNs) - m.preSystemCpuUsage
+
+		// Calculate cpu total usage percentage
+		cpuUsageTotal, err := event.GetValue("usage.total.ns")
```
Comment (ChrsMark) [inline metricbeat/module/containerd/cpu/metricset.go:126]: Can we put the rationale of these calculations into the PR's description and in the docs so as to have them clearly documented? I forsee getting questions about how these are calculated and we will have to dig into the code and struggle to understand these calculations.

File: metricbeat/module/containerd/cpu/metricset.go (hunk 3 above)
Comment (MichaelKatsoulis) [reply metricbeat/module/containerd/cpu/metricset.go:130]: remember it is a draft . I gave no intention of keeping it

File: metricbeat/module/containerd/cpu/metricset.go (hunk 5 above)
Comment (MichaelKatsoulis) [reply metricbeat/module/containerd/cpu/metricset.go:126]: Yes of course!

File: metricbeat/module/containerd/cpu/cpu.go (hunk 1 above)
Comment (MichaelKatsoulis) [reply metricbeat/module/containerd/cpu/cpu.go:53]: By mistake

File: metricbeat/module/containerd/cpu/metricset.go (hunk 4 above)
Comment (MichaelKatsoulis) [reply metricbeat/module/containerd/cpu/metricset.go]: Good idea

**Comment by MichaelKatsoulis** [comment]
Regarding the calculation of cpu usage percentage I followed this approach:

`Containerd` provides us with `container_cpu_total_nanoseconds`, `container_cpu_user_nanoseconds`, `container_cpu_kernel_nanoseconds` metrics per container id. 
And also `container_per_cpu_nanoseconds` for each container.

`container_cpu_total_nanoseconds` is the sum of the `container_per_cpu_nanoseconds ` for every cpu.

`Containerd` also provides `process_cpu_seconds_total` which is the `Total user and system CPU time spent in seconds.`

So in order to get the cpu usage percentage of each container I followed the approach we have in docker.

For each container:
`(container_cpu_total_nanoseconds - pre_container_cpu_total_nanoseconds) / (process_cpu_seconds_total - pre_process_cpu_seconds_total)`

In order to set the `pre_container_cpu_total_nanoseconds` and `pre_process_cpu_seconds_total` , every time new events are fetched , those values get updated with the latest values received.
The `pre_container_cpu_total_nanoseconds` is calculated per container id.

Basically, I took a point of reference, then see the difference in next batch of events. That way you can tell how much of the time was used by the container.

Also as the `container_cpu_total_nanoseconds` is the sum for all cpus , in order then to normalise the percentage , the percentage is divided with the number of cpus used by each container.

@fearful-symmetry as you have worked with docker module in similar calculations, what do you think about this approach?

File: x-pack/metricbeat/module/containerd/cpu/cpu.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+)
+
+const (
+	defaultScheme = "http"
+	defaultPath   = "/v1/metrics"
+)
+
+var (
+	// HostParser validates Prometheus URLs
+	hostParser = parse.URLHostParserBuilder{
```
Comment (fearful-symmetry) [inline x-pack/metricbeat/module/containerd/cpu/cpu.go:20]: Why is this declared globally? I only see it being used in the init function?

File: x-pack/metricbeat/module/containerd/cpu/cpu.go [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+		DefaultScheme: defaultScheme,
+		DefaultPath:   defaultPath,
+	}.Build()
+)
+
+// init registers the MetricSet with the central registry.
+// The New method will be called after the setup of the module and before starting to fetch data
+func init() {
+	// Mapping of state metrics
+	mapping := &prometheus.MetricsMapping{
```
Comment (fearful-symmetry) [inline x-pack/metricbeat/module/containerd/cpu/cpu.go:30]: Is there a reason why this map is declared here? I don't think that `getMetricsetFactory` function is getting reused, so we could just put it in one place?

File: x-pack/metricbeat/module/containerd/cpu/metricset.go [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+		if m.calcPct {
+			contCpus, ok := perContainerCpus[cID]
+			if !ok {
+				contCpus = 1
+			}
+			// calculate system usage delta
+			systemUsageDelta := float64(systemTotalNs) - m.preSystemCpuUsage
+
+			// Calculate cpu total usage percentage
+			cpuUsageTotal, err := event.GetValue("usage.total.ns")
```
Comment (fearful-symmetry) [inline x-pack/metricbeat/module/containerd/cpu/metricset.go:120]: This nearly identical logic block gets duplicated 3 times, we may want to try to abstract this away to the `calcCpuTotalUsagePct` function so it's a bit easier to wrangle.

File: x-pack/metricbeat/module/containerd/cpu/metricset.go [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+	//if percpu.(float64) != 0.0 {
+	//	if _, ok := perContainerCpus[cid]; ok {
+	//		perContainerCpus[cid] += 1
+	//	} else {
+	//		perContainerCpus[cid] = 1
+	//	}
+	//}
+}
+
+func calcCpuTotalUsagePct(cpuUsageTotal, systemUsageDelta, contCpus float64,
```
Comment (fearful-symmetry) [inline x-pack/metricbeat/module/containerd/cpu/metricset.go:194]: I'm not clear on why these are three separate functions. They look really similar?

File: x-pack/metricbeat/module/containerd/cpu/metricset.go [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+		}
+		// setting ECS container.id
+		containerFields := common.MapStr{}
+		var cID string
+		if containerID, ok := event["id"]; ok {
+			cID = (containerID).(string)
+			containerFields.Put("id", cID)
+			event.Delete("id")
+		}
+		e, err := util.CreateEvent(event, "containerd.cpu")
```
Comment (fearful-symmetry) [inline x-pack/metricbeat/module/containerd/cpu/metricset.go:95]: If we're gonna borrow functions from other modules, we should move that code outside the module to somewhere generic.

**Review by fearful-symmetry — CHANGES_REQUESTED** [review]

**Comment by fearful-symmetry** [comment]
As far as calculating the CPU percents, your logic seems sound?  Generally, we track a previous value, calculate the delta between the previous total and the current, and divide that by the time between the deltas. Pay attention to the values that are being reported upstream, as not all platforms will have a particularly clear-cut idea of what a "total" is.

Normalized CPU usage should be thought of as the total per-CPU. Put another way, the maximum value for `norm.pct` should be `100%`, and the maximum value for `pct` should usually be `100% * Number_of_cpus`

For example, check out https://github.com/elastic/beats/blob/master/metricbeat/module/docker/cpu/helper.go

and https://github.com/elastic/beats/blob/master/metricbeat/internal/metrics/cpu/metrics.go

**Comment by MichaelKatsoulis** [comment]
> Normalized CPU usage should be thought of as the total per-CPU. Put another way, the maximum value for norm.pct should be 100%, and the maximum value for pct should usually be 100% * Number_of_cpus

That was also my understanding. As `containerd` calculates the total cpu usage as the sum of the usage of each cpu, I understand that in order to normalise it, we need to divide with the number of cpus. If not, the percentage can be over 100%.

>Generally, we track a previous value, calculate the delta between the previous total and the current, and divide that by the time between the deltas.

Hmm. In order to get the percentage I expect a division with the total system usage delta in the same time and not the time between the deltas. That will be the way we get rates like in https://github.com/elastic/beats/blob/master/metricbeat/module/docker/diskio/helper.go#L171

**Comment by MichaelKatsoulis** [comment]
@fearful-symmetry Thanks for your comments. I restructured a bit the code following your suggestion. Feel free to have a look.

**Comment by fearful-symmetry** [comment]
> That will be the way we get rates like in https://github.com/elastic/beats/blob/master/metricbeat/module/docker/diskio/helper.go#L171

I may not be articulating myself very well, but that looks like more or less what I'm trying to say?

**Comment by fearful-symmetry** [comment]
Looking better. @MichaelKatsoulis do you want to add some unit or integration tests to the metricsets?

**Comment by MichaelKatsoulis** [comment]
> Looking better. @MichaelKatsoulis do you want to add some unit or integration tests to the metricsets?

Yes I will

**Comment by MichaelKatsoulis** [comment]
@ChrsMark , @fearful-symmetry this PR is ready for review now. Things missing are integrations tests which is a bit tricky in the `containerd` case and also a Kibana visualisation. They will be added in follow up prs.

File: x-pack/metricbeat/module/containerd/memory/memory.go [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+
+		// Calculate memory total usage percentage
+		if m.calcPct {
+			inactiveFiles, err := event.GetValue("inactiveFiles")
+			if err == nil {
+				usageTotal, err := event.GetValue("usage.total")
+				if err == nil {
+					memoryLimit, err := event.GetValue("usage.limit")
+					if err == nil {
+						// calculate working set memory usage
```
Comment (fearful-symmetry) [inline x-pack/metricbeat/module/containerd/memory/memory.go:126]: How confident are you That these files will all be the proper type, and that we can't get a division by zero here? It seems like every time I have an unprotected division operator in the system or linux module, I regret it.

Comment (MichaelKatsoulis) [reply x-pack/metricbeat/module/containerd/memory/memory.go:126]: I am confident that the type will always be float64 as containerd converts them to this type every time [(code)](https://github.com/containerd/containerd/blob/6fa958853197af31ffd1225a6c300589e5068667/metrics/cgroups/v1/memory.go#L568).
Regarding the zero division you are right, I will add a check

File: x-pack/metricbeat/module/containerd/helper.go [hunk 12]
```diff
@@ -0,0 +1,28 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package containerd
+
+import "github.com/elastic/beats/v7/libbeat/common"
+
+// SetCIDandNamespace sets container.id ECS field and containerd.namespace module field
+func SetCIDandNamespace(event common.MapStr) (common.MapStr, common.MapStr, string) {
```
Comment (ChrsMark) [inline x-pack/metricbeat/module/containerd/helper.go:10]: I would split it into 2 functions like `GetAndDeleteCid` and `GetAndDeleteNamespace`,  which would return the id and namespace explicitly. Current implementation gets an Event and returns `moduleFields`, `rootFields` and cID, so it's a mix and not a straightforward in<->out function.  Also "set" naming implies that the function sets sth on an already existent object, var or whatever.

File: x-pack/metricbeat/module/containerd/memory/_meta/data.json [hunk 13]
```diff
@@ -0,0 +1,19 @@
+{
```
Comment (ChrsMark) [inline x-pack/metricbeat/module/containerd/memory/_meta/data.json:1]: Same here, sample events must be as representative as possible since are supposed to provide good documentation.

File: x-pack/metricbeat/module/containerd/memory/memory.go [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+
+	for _, event := range events {
+
+		// setting ECS container.id and containerd.namespace
+		rootFields, moduleFields, cID := containerd.SetCIDandNamespace(event)
+
+		// Calculate memory total usage percentage
+		if m.calcPct {
+			inactiveFiles, err := event.GetValue("inactiveFiles")
+			if err == nil {
```
Comment (ChrsMark) [inline x-pack/metricbeat/module/containerd/memory/memory.go:121]: too much encapsulation here. You can just check if `err != nil` and if yes `continue`.

File: x-pack/metricbeat/module/containerd/blkio/_meta/docs.asciidoc [hunk 15]
```diff
@@ -0,0 +1 @@
+This is the blkio metricset of the module containerd.
```
Comment (ChrsMark) [inline x-pack/metricbeat/module/containerd/blkio/_meta/docs.asciidoc:1]: This docs page looks quite basic. Maybe we can cover the rational of how metrics and pcts are calculated here as suggested at https://github.com/elastic/beats/pull/29247#discussion_r761951922? The better we do our docs the easier it would be to provide support in the future.

File: x-pack/metricbeat/metricbeat.reference.yml [hunk 16]
```diff
@@ -427,6 +427,15 @@ metricbeat.modules:
   hosts: ["localhost:8500"]
 
 
+#------------------------------ Containerd Module ------------------------------
+- module: containerd
+  metricsets: ["cpu", "memory", "blkio"]
+  enabled: false
+  period: 10s
+  hosts: ["localhost:1338"]
```
Comment (ChrsMark) [inline x-pack/metricbeat/metricbeat.reference.yml:435]: Is this default endpoint of getting containerD metrics? In docker the default is a socket. I think we lack documentation on how/fromWhere one collect containerD metrics.

File: x-pack/metricbeat/metricbeat.reference.yml [hunk 17]
```diff
@@ -427,6 +427,15 @@ metricbeat.modules:
...
+  calcpct: true
```
Comment (ChrsMark) [inline x-pack/metricbeat/metricbeat.reference.yml:436]: This setting should be documented for better supportability.

File: x-pack/metricbeat/module/containerd/_meta/config.yml [hunk 18]
```diff
@@ -0,0 +1,7 @@
+- module: containerd
+  metricsets: ["cpu", "memory", "blkio"]
+  enabled: false
+  period: 10s
+  hosts: ["localhost:1338"]
+  calcpct: true
```
Comment (ChrsMark) [inline x-pack/metricbeat/module/containerd/_meta/config.yml:6]: Just a thought around this. Would it make sense to make it available per metricset instead of having it horizontal effect? Sth like:

```
calcpct.cpu: true
calcpct.memory: false
calcpctblkio: true
```

**Review by ChrsMark — COMMENTED** [review]
Looks good. Left some "prettify" comments about docs/settings mostly.

File: x-pack/metricbeat/module/containerd/helper.go (hunk 12 above)
Comment (MichaelKatsoulis) [reply x-pack/metricbeat/module/containerd/helper.go:10]: Ok!

File: x-pack/metricbeat/module/containerd/_meta/config.yml (hunk 18 above)
Comment (MichaelKatsoulis) [reply x-pack/metricbeat/module/containerd/_meta/config.yml:6]: For blkio it is not needed(we don't calculate any percentage) but for cpu and memory makes sense

Comment (MichaelKatsoulis) [reply x-pack/metricbeat/module/containerd/_meta/config.yml]: 5be9960ff9482e9b4b736f3fd742817293cf7c04

File: x-pack/metricbeat/module/containerd/helper.go (hunk 12 above)
Comment (MichaelKatsoulis) [reply x-pack/metricbeat/module/containerd/helper.go]: 5be9960ff9482e9b4b736f3fd742817293cf7c04

**Comment by MichaelKatsoulis** [comment]
@ChrsMark I updated the PR based on your comments. Could you take another look?

File: metricbeat/docs/modules/containerd.asciidoc [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+
+```
+[metrics]
+    address = "127.0.0.1:1338"
+```
+
+[float]
+=== Compatibility
+
+The Conatinerd module is tested with the following versions of Containerd:
```
Comment (hendry-lim) [inline metricbeat/docs/modules/containerd.asciidoc:37]:
```suggestion
The Containerd module is tested with the following versions of Containerd:
```

**Review by ChrsMark — APPROVED** [review]
lgtm

File: metricbeat/docs/modules/containerd.asciidoc [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+v1.5.2
+
+[float]
+=== Module-specific configuration notes
+
+For cpu metricset if `calcpct.cpu` setting is set to true, cpu usage percentages will be calculated
+and more specifically fields `containerd.cpu.usage.total.pct`, `containerd.cpu.usage.kernel.pct`, `containerd.cpu.usage.user.pct`.
+Default value is true.
+
+For memory metricset if `calcpct.memory` setting is set to true, memory usage percentages will be calculated
```
[the remark this answers is not in the document]
Comment (MichaelKatsoulis) [reply metricbeat/docs/modules/containerd.asciidoc:47]: The thought was initiated by @ChrsMark [comment](https://github.com/elastic/beats/pull/29247#discussion_r761951922). In general I also agree because those extra calculations (as well as extra iteration of the events because of them) may overload the system in case of too many containers. So it is safer to have it configurable. By default it is true anyway.

File: metricbeat/docs/modules/containerd.asciidoc [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+beta[]
+
+include::{libbeat-dir}/shared/integration-link.asciidoc[]
+
+:modulename!:
+
+Containerd module collects cpu, memory and blkio statistics about
+running containers controlled by containerd runtime.
+
+The current metricsets are: `cpu`, `blkio` and `memory`. They are not enabled by default.
```
[the remark this answers is not in the document]
Comment (MichaelKatsoulis) [reply metricbeat/docs/modules/containerd.asciidoc:20]: I am trying to understand how this text in blkio.asciidoc  is generated

[the remark this answers is not in the document]
Comment (MichaelKatsoulis) [reply metricbeat/docs/modules/containerd.asciidoc]: Ok this comes from this [line](https://github.com/elastic/beats/pull/29247/files#diff-355bdb78bafbb3428116b7e15a12093cedbceb9dfb2d00eaec9437612d00a9a4R73).
I believe that all these metricests should be enabled by default when a user enables containerd module. We do that in most of the modules. So I made an update in 69f4b5f5b5803a8ac685cf82e977818711fe5b14

File: x-pack/metricbeat/module/containerd/memory/memory.go [hunk 22]
```diff
@@ -0,0 +1,161 @@
...
+
+		containerFields.Put("id", cID)
+		rootFields.Put("container", containerFields)
+		moduleFields.Put("namespace", namespace)
+
+		// Calculate memory total usage percentage
+		if m.calcPct {
+			inactiveFiles, err := event.GetValue("inactiveFiles")
+			if err != nil {
+				continue
```
Comment (tetianakravchenko) [inline x-pack/metricbeat/module/containerd/memory/memory.go:133]: in which cases `inactiveFiles` can be not present in event? should here be added a debug log to explain why usage percentage was skipped?

File: x-pack/metricbeat/module/containerd/cpu/cpu.go [hunk 23]
```diff
@@ -0,0 +1,234 @@
...
+				contCpus = 1
+			}
+			// calculate timestamp delta
+			timestampDelta := int64(0)
+			if !m.preTimestamp.IsZero() {
+				timestampDelta = timestamp.UnixNano() - m.preTimestamp.UnixNano()
+			}
+			// Calculate cpu total usage percentage
+			cpuUsageTotal, err := event.GetValue("usage.total.ns")
+			if err == nil {
```
[the remark this answers is not in the document]
Comment (MichaelKatsoulis) [reply x-pack/metricbeat/module/containerd/cpu/cpu.go:149]: In this case, we iterate a batch of events and we check if `usage.total.ns` field is present. There is for sure one event in this batch that this field is not present and that is the one that has `system.total` field. The reason is that `process_cpu_seconds_total` containerd metric that is then mapped to `system.total`  field, does not include info about any container id (system wise metric not container specific). So that leads to an event that has only `system.total` field.  While the rest of the events have fields that are grouped together due to `container_id` mainly. So with this `if` I just want to be sure that we skip trying to calculate percentages for the event that only has `system.total` (it does not include any other field, is not container specific). It is not an actual error to log anything. But I will add a comment.

File: x-pack/metricbeat/module/containerd/memory/memory.go (hunk 22 above)
Comment (MichaelKatsoulis) [reply x-pack/metricbeat/module/containerd/memory/memory.go:133]: In no case. Only if there is an error. I will add a debug

**Review by tetianakravchenko — APPROVED** [review]

**Comment by MichaelKatsoulis** [comment]
@fearful-symmetry you have some requested changes that are blocking the merging of the PR. I can still merge it though,  unless you want to make a final review

**Comment by fearful-symmetry** [comment]
@MichaelKatsoulis Sorry about that! No idea how I missed your ping last week.

**Comment by MichaelKatsoulis** [comment]
> @MichaelKatsoulis Sorry about that! No idea how I missed your ping last week.
No problem @fearful-symmetry . You can still give it a look and if there is something you think is important I can open a follow up pr to fix it.
